examples/asr/speech_to_text_streaming_infer.py

```diff
@@ -0,0 +1,293 @@
+
+"""
+
+
+import contextlib
+import json
+import time
+from argparse import ArgumentParser
+
+import torch
+from omegaconf import open_dict
+
+import nemo.collections.asr as nemo_asr
+from nemo.collections.asr.metrics.wer import word_error_rate
+from nemo.collections.asr.parts.utils.rnnt_utils import Hypothesis
+from nemo.collections.asr.parts.utils.streaming_utils import FramewiseStreamingAudioBuffer
+from nemo.utils import logging
+
+
+def extract_transcribtions(hyps):
+    if isinstance(hyps[0], Hypothesis):
+        transcribtions = []
+        for hyp in hyps:
+            transcribtions.append(hyp.text)
+    else:
+        transcribtions = hyps
+    return transcribtions
+
+
+def calc_drop_extra_pre_encoded(asr_model, step_num):
+    # for the first step there is no need to drop any tokens after the downsampling as no caching is being used
+    if step_num == 0:
+        return 0
+    else:
+        return asr_model.encoder.streaming_cfg.drop_extra_pre_encoded
+
+
+def perform_streaming(asr_model, streaming_buffer, compare_vs_offline=False, debug_mode=False):
+    batch_size = len(streaming_buffer.streams_length)
+    if compare_vs_offline:
+        # would run the whole audio at once through the model like offline mode in order to compare the results with the stremaing mode
+        # the output of the model in the offline and streaming mode should be exactly the same
+        with autocast():
+            processed_signal, processed_signal_length = streaming_buffer.get_all_audios()
+            with torch.no_grad():
+                (
+                    pred_out_offline,
+                    transcribed_texts,
+                    cache_last_channel_next,
+                    cache_last_time_next,
+                    best_hyp,
+                ) = asr_model.stream_step(
+                    processed_signal=processed_signal,
+                    processed_signal_length=processed_signal_length,
+                    return_transcribtion=True,
+                )
+        final_offline_tran = extract_transcribtions(transcribed_texts)
+        logging.info(f" Final offline transcriptions:   {final_offline_tran}")
+    else:
+        final_offline_tran = None
+
+    cache_last_channel, cache_last_time = asr_model.encoder.get_initial_cache_state(batch_size=batch_size)
+
+    previous_hypotheses = None
+    streaming_buffer_iter = iter(streaming_buffer)
+    pred_out_stream = None
+    for step_num, (chunk_audio, chunk_lengths) in enumerate(streaming_buffer_iter):
+        with autocast():
+            with torch.no_grad():
+                (
+                    pred_out_stream,
+                    transcribed_texts,
+                    cache_last_channel,
+                    cache_last_time,
+                    previous_hypotheses,
+                ) = asr_model.stream_step(
```

Comment:
Not sure if chunkwise is the best term here as it is a very general term which can be used for any asr inference. Also it can be confusing with our chunked inference. These are real streaming mode and stream looks to be a more accurate term here. Actually I just checked Lingvo and they used similar term:

https://github.com/tensorflow/lingvo/blob/d2f1e1b3cccdac8f73ae20f86afb03560b1c176d/lingvo/core/conformer_layer.py#L387